File: programming_language/javasolutions/p79/word-search.py
class WordSearch:
    def exist(self, board, word):
        if not word:
            return True

        if not board or not board[0]:
            return False

        m = len(board)
        n = len(board[0])
        visited = [[False for _ in range(n)]
                   for _ in range(m)]
        
        for i in range(m):
            for j in range(n):
                if self.dfs(i, j, 0, word, visited, board):
                    return True
        return False

    def dfs(self, x, y, idx, word, visited, board):
        if idx == len(word):
            return True

        if x < 0 or x >= len(board) or y < 0 or y >= len(board[0]) or visited[x][y] or board[x][y] != word[idx]:
            return False

        visited[x][y] = True
        dx = [1, 0, -1, 0]
        dy = [0, 1, 0, -1]
        answer = False
        for i in range(4):
            xx = x + dx[i]
            yy = y + dy[i]
            if self.dfs(xx, yy, idx + 1, visited, board):
                answer = True
                break

        visited[x][y] = False
        return answer

File: programming_language/javasolutions/p79/WordSearch.java
package javasolutions.p79;

public class WordSearch {
  public boolean exist(char[][] board, String word) {
    if (word.isEmpty()) return true;
    if (board.length == 0 || board[0].length == 0) return false;

    boolean[][] visited = new boolean[board.length][board[0].length];

    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[0].length; j++) {
        if (existDFS(board, visited, 0, word, i, j)) return true;
      }
    }

    return false;
  }

  private boolean existDFS(char[][] board, boolean[][] visited, int index, String word, int row, int col) {
    if (index == word.length()) return true;
    if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || visited[row][col] || board[row][col] !=  word.charAt(index)) return false;

    visited[row][col] = true;
    boolean answer = existDFS(board, visited, index + 1, word, row - 1, col)
            || existDFS(board, visited, index + 1, word, row + 1, col)
            || existDFS(board, visited, index + 1, word, row, col - 1)
            || existDFS(board, visited, index + 1, word, row, col + 1);
    visited[row][col] = false;

    return answer;
  }
}